src/main/java/com/internet/shop/dao/jdbc/ShoppingCartDaoJdbcImpl.java

```diff
@@ -0,0 +1,180 @@
+import com.internet.shop.lib.Dao;
+import com.internet.shop.model.Product;
+import com.internet.shop.model.ShoppingCart;
+import com.internet.shop.util.ConnectionUtil;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@Dao
+public class ShoppingCartDaoJdbcImpl implements ShoppingCartDao {
+    @Override
+    public Optional<ShoppingCart> getByUserId(Long userId) {
+        String query = "SELECT * FROM shopping_carts WHERE user_id = ? AND deleted = FALSE";
+        Optional<ShoppingCart> searchedCart = Optional.empty();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, userId);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                searchedCart = Optional.of(getCartFromResultSet(resultSet));
+            }
+            if (searchedCart.isEmpty()) {
+                return searchedCart;
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Couldn't find cart by userId: "
+                    + userId, e);
+        }
+        searchedCart.get().setProducts(getProductsListByCartId(searchedCart.get().getId()));
+        return searchedCart;
+    }
+
+    @Override
+    public ShoppingCart create(ShoppingCart cart) {
+        String query = "INSERT INTO shopping_carts (user_id) VALUES (?)";
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection
+                    .prepareStatement(query, Statement.RETURN_GENERATED_KEYS);
+            statement.setLong(1, cart.getUserId());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                cart.setId(resultSet.getLong(1));
+            }
+            return cart;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Couldn't create provided cart: " + cart, e);
+        }
+    }
+
+    @Override
+    public Optional<ShoppingCart> getById(Long cartId) {
+        String query = "SELECT * FROM shopping_carts WHERE cart_id = ? AND deleted = FALSE";
+        List<Product> productsInCart = getProductsListByCartId(cartId);
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(query);
+            statement.setLong(1, cartId);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                ShoppingCart searchedCart = getCartFromResultSet(resultSet);
+                searchedCart.setProducts(productsInCart);
```

Comment:
Inner connection

Reply:
there is no nested connection, it was created and closed in method getProductsListByCartId, that is located above this TWR statement

Reply:
Sorry, misunderstanding
Thought it was a private method
🙅‍♀️